### PR TITLE
[FIX] Fix the hang in cuGraph Python Uniform Neighbor Sample, Add Logging to Bulk Sampler

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Some dask operations were not being done correctly, and time was being lost in broadcasting the rank and label arrays to all workers.  This PR resolves those issues.

Also pulls in the previously-experimental changes that add logging to the bulk sampler.

Credit to @VibhuJawa for isolating and fixing the issues with the column merge in `uniform_neighbor_sample` and the new sampling notebook and shell script.

This PR does modify the sampling APIs so it is breaking.  The API changes are necessary to avoid unnecessary shuffling, and eventually, to improve batch id assignment.

**Dataset:** `ogbn_papers100M x 2`; **Fanout:** `[25, 25]`; **Batch Size:** `512`; **Seeds Per Call:** `524288`
Current runtime: 2.69 s ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)
Previous runtime: 4.51 s ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)
Speedup: 1.7x

**Dataset:** `ogbn_papers100M x 4`; **Fanout:** `[25, 25]`; **Batch Size:** `512`; **Seeds Per Call:** `524288`
Current runtime: 6.32 s ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)
Previous runtime: 10.7 s ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)
Speedup: 1.7x

### CODE REVIEW

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:264]: Why did we switch to `client.compute` here from `client.submit `?

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:472]: I think we should use  `isinstance` to be cleaner. 
```suggestion
    if isinstance(ddf, cudf.DataFrame):
```

File: python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py:202]:
```suggestion
        if isinstace(batch_id_filter, dask_cudf.DataFrame):
```

File: python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py:251]:
```suggestion
        if isinstance(self.__batches, dask_cudf.DataFrame):
```

File: python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py:258]: This time is not accurate as we are not waiting here as persist is not a sync operation. I would probably just remove this.

File: python/cugraph/cugraph/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/sampling/uniform_neighbor_sample.py:187]: Recommend checking type of the start_list as cudf.Series.

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Requested some minor changes .  Looks good overall .

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (alexbarghi-nv) [reply python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:264]: I did that based off your suggestion.  It fixed the hang between calling `uniform_neighbor_sample` and writing the parquet files.

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:240]: I suggest keeping this as we are releasing the futures below. It can potentially cause issues .

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:245]: Same here

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Some more minor changes, but i think we should be good to go after that.  I think the only thing is what @rlratzel requested around keeping the now legacy implementation around.

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [reply python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:240]: Suggest using wait like this. 

```python3
wait([ddf, ddf_offsets])
```

**Comment by VibhuJawa** [comment]
This still seems to be actually **hanging** (not perf issues) for me when running 10 times on bulk samplinng. :-(  . Can you try running it and ensure we dont cause hangs. 


```python3
dataset='ogbn_papers100M'
dataset_root="/datasets/abarghi/"
output_root="/raid/vjawa/"
reverse_edges=True
add_edge_types=False
batch_size=1024
seeds_per_call=524288
fanout=[10,10,10]
replication_factor=4
seed=123

dataset_dir=dataset_root
output_path=output_root
persist=False
```

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:679]: We probably need to raise a depracted warning here.

File: python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py:259]: I gather this is us being extra careful to prevent problems

File: python/cugraph/cugraph/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/sampling/uniform_neighbor_sample.py:288]: Same ask here to raise a depractated warning.

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Minor asks around raising deprecated warnings

File: python/cugraph/cugraph/sampling/uniform_neighbor_sample.py
Comment (alexbarghi-nv) [reply python/cugraph/cugraph/sampling/uniform_neighbor_sample.py:288]: it should raise a warning.  That function is wrapped with the deprecated warning wrapper

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (alexbarghi-nv) [reply python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:679]: it should raise a warning. That function is wrapped with the deprecated warning wrapper

File: python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py
Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py:259]: yeah

File: python/cugraph/cugraph/sampling/uniform_neighbor_sample.py
Comment (alexbarghi-nv) [inline python/cugraph/cugraph/sampling/uniform_neighbor_sample.py:180]: @VibhuJawa deprecated wrapper is here

File: python/cugraph/cugraph/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [reply python/cugraph/cugraph/sampling/uniform_neighbor_sample.py:288]: Ahh missed that.

**Review by VibhuJawa — APPROVED** [review]
LGTM

File: python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py
Comment (alexbarghi-nv) [inline python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py:541]: @VibhuJawa deprecated wrapper is here

File: python/cugraph/cugraph/sampling/uniform_neighbor_sample.py
Comment (VibhuJawa) [reply python/cugraph/cugraph/sampling/uniform_neighbor_sample.py:180]: Thanks @alexbarghi-nv

File: python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: Why did we add this compute back ?

Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: `client.submit` doesn't accept it uncomputed.  What we had before worked with `client.compute`.  I'm trying to find alternative that works for `submit`.

The error we get is this:
```
AttributeError("'Scalar' object has no attribute '_parent_meta'")
```

And wrapping with delayed doesn't work either, it gives a similar error.

Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: i.e.
`delayed(lambda df : df.min())(self.__batches)` doesn't work

Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: I think I found a way around that issue, but we have another problem where the execution hangs because we are setting allow_other_workers=False so it can't get the min

Comment (VibhuJawa) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: I guess we can let it be for now and try to get other optimizations as a followup

Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: I've since figured out that what we had with `client.compute` and not computing the min beforehand was wrong.  It should cause a hang, which we were seeing.

Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py]: #3690

File: benchmarks/cugraph/standalone/bulk_sampling/benchmarking_script.ipynb
Comment (rlratzel) [inline benchmarks/cugraph/standalone/bulk_sampling/benchmarking_script.ipynb:1]: A notebook is nice, but just so I can plan ahead: do we want to run this notebook as part of the nightlies, and if so, can we use the notebook conversion script to run it from a command-line (this is what we do in CI, as seen [here](https://github.com/rapidsai/cugraph/blob/branch-23.08/ci/test_notebooks.sh), which calls [this](https://github.com/rapidsai/cugraph/blob/branch-23.08/ci/utils/nbtest.sh))

File: benchmarks/cugraph/standalone/bulk_sampling/bulk_sampling.sh
Comment (rlratzel) [inline benchmarks/cugraph/standalone/bulk_sampling/bulk_sampling.sh:20]: I'm guessing at a minimum we'll want these as params to the script?  a `FIXME:` is fine for now IMO.

**Review by rlratzel — APPROVED** [review]
LGTM, and thanks for adding support for the legacy args.  I'm filing [this issue](https://github.com/rapidsai/cugraph/issues/3698) to remind us to remove it for 23.10 (which gives users one release to migrate).

I have a question and a possible FIXME suggestion which need not hold up approval now.

File: benchmarks/cugraph/standalone/bulk_sampling/bulk_sampling.sh
Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/bulk_sampling/bulk_sampling.sh]: resolved

File: benchmarks/cugraph/standalone/bulk_sampling/benchmarking_script.ipynb
Comment (alexbarghi-nv) [reply benchmarks/cugraph/standalone/bulk_sampling/benchmarking_script.ipynb:1]: I think we do want to run this nightly.

**Pull request merged**
